### DISCUSSION TITLE
❗ TSOL-MS2000 now enforces SSL. Use Proxy Vers >= 0.10.0 ❗

### DISCUSSION THREAD

**Discussion by s-allius** [s-allius/tsun-gen3-proxy#113] [Announcements]
In June, I and a lot of other users got an upgrade from TSUN, which breaks the compatibility with this proxy.

## In details:
With the upgrade, TSUN installs a firmware that only supports connections via SSL. You can recognize this by the fact that you see a connection from your inverter to port 10443 in your router. The web configuration of the inverter is still accessible at `http://10.10.100.254/` and so in principle it is possible to change the firmware and the configuration for an unencrypted connection to the proxy. At present, operation is no longer possible with this proxy after the update to the SSL firmware! 

## Solution: Client-Mode
If you have firmware with SSL support on your GEN3PLUS inverter, you can use the new `client mode` to retrieve data from your inverter. This was introduced with proxy version 0.10.

See also: [Operation-Modes-Overview/client-mode](https://github.com/s-allius/tsun-gen3-proxy/wiki/Operation-Modes-Overview#client-mode)

## How to avoid an update:
ADD this line, or change the `tsun.allow` statement to:
```bash
tsun.allow = ['AT+Z']
```

the default statement: 
```bash
tsun.allow = ['AT+Z', 'AT+UPURL', 'AT+SUPDATE']   # this allow updates an can brake your setup
```
allows updates of the DSP (`AT+SUPDATE`) and the logger (`AT+UPURL`) firmware, which can break your setup

You find a description of most of the AT commands here: [AT-commands](https://github.com/s-allius/tsun-gen3-proxy/wiki/AT--commands)
and how to block them here: [AT-command-filter](https://github.com/s-allius/tsun-gen3-proxy/wiki/Configuration-details#at-command-filter-for-gen3plus-inverter-only)

**Comment by bric** [comment]
to bad, I saw the 10443 did not read your thread and changed it back to 10000. Now the MS2000 is no longer accessible. Do you have a hint how to bring it back to factory default ? 
Thank you, and also a big thank you for the stable proxy implementation ...
Andreas

**Reply by bric** [reply]
so, got it back online somehow by blocking the wlan connection on my router, reboot the ms2000, reconnect via AP, change port back to 10443. Re-enable wlan. 
Andreas

**Reply by s-allius** [reply]
I have the same problem. The inverter don't establish a connection to any port. The Web-Config isn't reachable anymore. The AP_ SSID is working, but there is also no open port for a web-config :-(

**Reply by s-allius** [reply]
@bric: It works again thank you. I disabled my WiFi, switched off the Inverter Grid connection for a few second. An after some minutes I could connect to the inverters AP_ SSID and have access to the Web-Config again.

Thank you very much, for you hint

**Comment by s-allius** [comment]
Today I make some experiences with TLS. But unfortunately the inverter checks the certificate and don't like my self-signed certificate, since it was not from the right CA.

**Comment by Timurling** [comment]
Expiriencing the same issue here.

Edit:
Happy to support with testing something incase it helps

I have also notice that everytime I change from Domain to (local) IP in the settings it gets set back to Domain with iot.talent-monitoring.com after reboot.

**Comment by s-allius** [comment]
Good news!
I have just established an unencrypted connection from the proxy to the inverter and received some measured values. So we don't need to downgrade to an older firmware without SSL support.
I still have to think about the SW design. But I'm pretty sure that the next proxy version will also support GEN3PLUS inverters with SSL firmware again.

**Reply by Timurling** [reply]
Thats fantastic! Thanks for putting all this effort into such a great tool!

**Comment by s-allius** [comment]
Today I build a preview build of version 0.10.0 for testing the connection over port 8899.

you can get the image with the label preview:
```bash
docker pull ghcr.io/s-allius/tsun-gen3-proxy:0.10.0-preview.1
```
and you have to adapt your configuration: https://github.com/s-allius/tsun-gen3-proxy/issues/117#issuecomment-2214750886

Any feedback is warmly welcome.

**Reply by Timurling** [reply]
Just installed it. 
I think its important to mention to add port 8899 to the compose file.

Will run it for the next day and provide feedback.

**Reply by s-allius** [reply]
Because it is an outgoing connection you don't need to add it to the compose file. The compose file only open ports for incoming connections.

**Reply by Timurling** [reply]
Right!

Its running now for several hours and I can pull the data into ioBroker using mqtt. I can also control the rated power. No issues so far! However its interresting that the TSUN App shows a lower "Total Generation" then the inverter provides via mqtt.
Thanks a lot for your work!

**Reply by Thunder8Storm** [reply]
working here too

**Reply by Timurling** [reply]
@s-allius 

This morning I see multiple times the following Error on the Logs:

'2024-07-10 05:09:23 ERROR | root | ModbusTcpCreate: Exception for ('192.168.0.34', 8899):
Traceback (most recent call last):
  File "/home/tsun-proxy/modbus_tcp.py", line 53, in modbus_loop
    async with ModbusConn(host, port) as stream:
  File "/home/tsun-proxy/modbus_tcp.py", line 22, in __aenter__
    reader, writer = await connection
                     ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/streams.py", line 48, in open_connection
    transport, _ = await loop.create_connection(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1122, in create_connection
    raise exceptions[0]
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1104, in create_connection
    sock = await self._connect_sock(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 1007, in _connect_sock
    await self.sock_connect(sock, address)
  File "/usr/local/lib/python3.12/asyncio/selector_events.py", line 651, in sock_connect
    return await fut
           ^^^^^^^^^
  File "/usr/local/lib/python3.12/asyncio/selector_events.py", line 691, in _sock_connect_cb
    raise OSError(err, f'Connect call failed {address}')
OSError: [Errno 113] Connect call failed ('192.168.0.34', 8899)

**Reply by s-allius** [reply]
Thank you for the report. I improved the error handling for the client-mode and build a new `preview` version. 
```
docker pull ghcr.io/s-allius/tsun-gen3-proxy:0.10.0-preview.2
```

**Reply by Timurling** [reply]
@s-allius 

When I first time run preview.2, without correct config.toml and mqtt configuration is throws the below error. Once I placed config.toml and restartet the container it did no appear again. Will run the new version for a couple of hours and provide feedback.

'2024-07-11 08:17:44  INFO | root | Event loop is stopped'
'2024-07-11 08:17:44  INFO | root | Close MQTT Task'
Traceback (most recent call last):
  File "/home/tsun-proxy/mqtt.py", line 69, in __loop
    async with self.__client:
  File "/usr/local/lib/python3.12/site-packages/aiomqtt/client.py", line 703, in __aenter__
    raise MqttError(str(exc)) from None
aiomqtt.exceptions.MqttError: [Errno -2] Name does not resolve
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "/home/tsun-proxy/./server.py", line 205, in <module>
    Inverter.class_close(loop)
  File "/home/tsun-proxy/inverter.py", line 78, in class_close
    loop.run_until_complete(cls.mqtt.close())
  File "/usr/local/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/home/tsun-proxy/mqtt.py", line 40, in close
    await self.task
  File "/home/tsun-proxy/mqtt.py", line 118, in __loop
    await asyncio.sleep(interval)
  File "/usr/local/lib/python3.12/asyncio/tasks.py", line 665, in sleep
    return await future
           ^^^^^^^^^^^^
asyncio.exceptions.CancelledError

**Comment by s-allius** [comment]
If you have firmware with SSL support on your GEN3PLUS inverter, you can use the new 'client mode' to retrieve data from your inverter. This was introduced with proxy version 0.10.

See also: [Operation-Modes-Overview/client-mode](https://github.com/s-allius/tsun-gen3-proxy/wiki/Operation-Modes-Overview#client-mode)
